### ISSUE TITLE
Eval imported code asynchronously

### ISSUE THREAD

**Issue by jjrv** [systemjs/systemjs#1223]
Chrome seems to disregard source maps for code that runs inside [this `eval()`](https://github.com/systemjs/systemjs/blob/13329572a31225298178b337569de1ec921536a4/lib/global-eval.js#L73) immediately when a module is imported synchronously (see #511). If there was a way to hook that call and run my own `eval()` function returning a promise, it could wrap the contents in `setImmediate()` and add a call to resolve the promise at the end. Then the actual code would execute outside the native `eval()` function, hopefully fixing the source map problem.

SystemJS would of course have to wait for the promises of all imported modules to resolve, before running the code that imported them, but it's already waiting for the XHR promises to download such code.

The hook would also allow moving any `//# sourceMappingURL` comments in the original imported code, to the last line also in the code that actually gets evaluated. Currently no hook seems to have a chance to run between [this line](https://github.com/systemjs/systemjs/blob/13329572a31225298178b337569de1ec921536a4/lib/cjs.js#L121) and the `eval()`.

**Comment by jjrv** [comment]
Seems I may have made a mistake and can get this working in client-side code. Sorry for the inconvenience, closing for now.

**Comment by jjrv** [comment]
Turns out at least for CommonJS it's quite easy to separate evaluating imported code and later running it on a different tick, but it doesn't fix the source map issue (there's another fix for it in #511 now).

Async eval code:

```JavaScript
System.parsedTbl = {};

var oldInstantiate = System.instantiate;
System.instantiate = function(load) {
	return(oldInstantiate.call(this, load).then(function(result) {
		if(!load.metadata.bundle && load.metadata.format == 'cjs') {
			var argList = [
				'__cjsWrapper',
				'require',
				'exports',
				'module',
				'__filename',
				'__dirname',
				'global',
				'GLOBAL'
			].concat(Object.keys(load.metadata.globals || {})).join(', ');

			eval('System.parsedTbl["' + load.name + '"] = function(' + argList + ') {' + load.source + '\n};');
			load.source = 'System.parsedTbl["' + load.name + '"](' + argList + ');';
			return(result);
		}

		return(result);
	}));
}
```
